optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/common/ConstraintStreamHelper.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.common;
+
+interface ConstraintStreamHelper<Right, JoinedStream, Joiner, Predicate> {
```

Comment:
I expect a lively discussion about this. :-)

This exists so that I can reuse `JoinerUtils.join(...)` across all the streams. It may be a lot of code for such a simple purpose, but it's just a level of abstraction. The code that is being shared thanks to this is complicated and needs to be kept identical across all the streams, so I think it's worth it.